File: src/coordination/__tests__/AsyncTimerEvent.test.ts
// @leosingleton/commonlibs - Common Libraries for TypeScript and .NET Core
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

import { AsyncTimerEvent } from '../AsyncTimerEvent';
import { Task } from '../../dotnet/Task';

describe('AsyncTimerEvent', () => {

  it('Tests the timer without repeat enabled', async () => {
    let timer = new AsyncTimerEvent(1000);
    let hasFired = false;
    setTimeout(async () => {
      await timer.waitAsync();
      hasFired = true;
    });

    await Task.delay(900);
    expect(hasFired).toBeFalsy();

    await Task.delay(200);
    expect(hasFired).toBeTruthy();
  });

  it('Tests the timer with repeat enabled', async () => {
    let timer = new AsyncTimerEvent(100, true);
    let fireCount = 0;
    setTimeout(async () => {
      while (fireCount < 12) {
        await timer.waitAsync();
        fireCount++;
      }
    });

    await Task.delay(1000);
    expect(fireCount).toBeGreaterThanOrEqual(8);
    expect(fireCount).toBeLessThanOrEqual(12);
    await Task.delay(500);
  });
});

File: dotnet/UnitTests/Coordination/AsyncTimerEventTest.cs
// LeoSingleton.CommonLibs - Common Libraries for TypeScript and .NET Core
// Copyright (c) Leo C. Singleton IV <leo@leosingleton.com>
// See LICENSE in the project root for license information.

using System.Threading.Tasks;
using Xunit;

namespace LeoSingleton.CommonLibs.Coordination.UnitTests
{
    public class AsyncTimerEventTest
    {
        /// <summary>
        /// Tests the timer without repeat enabled
        /// </summary>
        [Fact]
        public async Task SingleFire()
        {
            var timer = new AsyncTimerEvent(1000);
            bool hasFired = false;
            var unused = Task.Run(async () =>
            {
                await timer.WaitAsync();
                hasFired = true;
            });

            await Task.Delay(900);
            Assert.False(hasFired);

            await Task.Delay(200);
            Assert.True(hasFired);
        }

        /// <summary>
        /// Tests the timer with repeat enabled
        /// </summary>
        [Fact]
        public async Task MultipleFire()
        {
            var timer = new AsyncTimerEvent(100, true);
            int fireCount = 0;
            var unused = Task.Run(async () =>
            {
                while (fireCount < 12)
                {
                    await timer.WaitAsync();
                    fireCount++;
                }
            });

            await Task.Delay(1000);
            Assert.InRange(fireCount, 8, 12);
            await Task.Delay(500);
        }
    }
}
